Index: engine.c
===================================================================
RCS file: /build/data/openbsd/cvs/src/usr.bin/make/engine.c,v
diff -u -p -r1.74 engine.c
--- engine.c	9 Apr 2024 15:08:21 -0000	1.74
+++ engine.c	30 Jan 2025 14:34:38 -0000
@@ -579,8 +579,6 @@ handle_job_status(Job *job, int status)
 			 * JOB_IS_EXPENSIVE, perform the computation for
 			 * sequential make to figure out whether to display the
 			 * command or not.  */
-			if ((job->flags & JOB_SILENT) && sequential)
-				determine_expensive_job(job);
 			if ((job->flags & (JOB_SILENT | JOB_IS_EXPENSIVE)) 
 			    == JOB_SILENT)
 				printf(": %s", job->cmd);
Index: job.c
===================================================================
RCS file: /build/data/openbsd/cvs/src/usr.bin/make/job.c,v
diff -u -p -r1.166 job.c
--- job.c	18 Jun 2024 02:11:03 -0000	1.166
+++ job.c	30 Jan 2025 18:53:29 -0000
@@ -93,6 +93,10 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <sys/poll.h>
+#include <netdb.h>
 #include "defines.h"
 #include "job.h"
 #include "engine.h"
@@ -115,7 +119,6 @@ static int	aborting = 0;	    /* why is t
 static bool	no_new_jobs;	/* Mark recursive shit so we shouldn't start
 				 * something else at the same time
 				 */
-bool sequential;
 Job *runningJobs;		/* Jobs currently running a process */
 Job *errorJobs;			/* Jobs in error at end */
 Job *availableJobs;		/* Pool of available jobs */
@@ -123,6 +126,12 @@ static Job *heldJobs;		/* Jobs not runni
 static pid_t mypid;		/* Used for printing debugging messages */
 static Job *extra_job;		/* Needed for .INTERRUPT */
 
+/* to adjust the number of jobs dynamically */
+static int control_point = -1;
+struct pollfd pollarray[1];
+int jobs_goal;
+int allocated;
+
 static volatile sig_atomic_t got_fatal;
 
 static volatile sig_atomic_t got_SIGINT, got_SIGHUP, got_SIGQUIT, got_SIGTERM, 
@@ -151,6 +160,13 @@ static void may_remove_target(Job *);
 static void print_error(Job *);
 static void internal_print_errors(void);
 
+/* functions that adjust the number of jobs */
+static void adjust_up(void);
+static void may_adjust_jobs_goal();
+static int connect_to_local_addr(const char *);
+static int connect_to_inet(const char *, const char *);
+static int connect_to(char *);
+
 static int dying_signal = 0;
 
 const char *	basedirectory = NULL;
@@ -273,6 +289,7 @@ print_error(Job *j)
 	Buf_Destroy(&buf);
 	free(j->cmd);
 }
+
 static void
 quick_summary(int signo)
 {
@@ -540,8 +557,14 @@ postprocess_job(Job *job)
 		job->next = errorJobs;
 		errorJobs = job;
 	} else {
-		job->next = availableJobs;
-		availableJobs = job;
+		if (allocated <= jobs_goal) {
+			job->next = availableJobs;
+			availableJobs = job;
+	    	} else {
+			free(job);
+			allocated--;
+			may_adjust_jobs_goal();
+		}
 	}
 
 	if (errorJobs != NULL && aborting != ABORT_INTERRUPT)
@@ -658,7 +681,7 @@ may_continue_job(Job *job)
 		bool finished = job_run_next(job);
 		if (finished)
 			postprocess_job(job);
-		else if (!sequential)
+		else 
 			determine_expensive_job(job);
 	}
 }
@@ -753,6 +776,8 @@ reap_jobs(void)
 	bool reaped = false;
 	Job *job;
 
+	if (control_point != -1)
+		may_adjust_jobs_goal();
 	while ((pid = waitpid(WAIT_ANY, &status, WNOHANG)) > 0) {
 		if (WIFSTOPPED(status))
 			continue;
@@ -817,26 +842,25 @@ loop_handle_running_jobs(void)
 		handle_running_jobs();
 }
 
+
+
 void
 Job_Init(int maxJobs)
 {
-	Job *j;
-	int i;
-
 	runningJobs = NULL;
 	heldJobs = NULL;
 	errorJobs = NULL;
 	availableJobs = NULL;
-	sequential = maxJobs == 1;
+	jobs_goal = maxJobs;
 
+	if (control_point != -1)
+		may_adjust_jobs_goal();
+
+	adjust_up();
 	/* we allocate n+1 jobs, since we may need an extra job for
 	 * running .INTERRUPT.  */
-	j = ereallocarray(NULL, sizeof(Job), maxJobs+1);
-	for (i = 0; i != maxJobs; i++) {
-		j[i].next = availableJobs;
-		availableJobs = &j[i];
-	}
-	extra_job = &j[maxJobs];
+	/* we need the extra_job for running .INTERRUPT */
+	extra_job = emalloc(sizeof(Job));
 	mypid = getpid();
 
 	aborting = 0;
@@ -846,6 +870,8 @@ Job_Init(int maxJobs)
 bool
 can_start_job(void)
 {
+	if (control_point != -1)
+		may_adjust_jobs_goal();
 	if (aborting || availableJobs == NULL)
 		return false;
 	else
@@ -956,4 +982,173 @@ Job_AbortAll(void)
 	 */
 	while (waitpid(WAIT_ANY, &foo, WNOHANG) > 0)
 		continue;
+}
+
+/* Job control protocol: if the right environment variables are defined,
+ * the actual number of parallel jobs may be adjusted from outside.
+ */
+static void
+adjust_up(void)
+{
+	Job *j;
+
+	for (;allocated < jobs_goal; allocated++) {
+		j = emalloc(sizeof(Job));
+		j->next = availableJobs;
+		availableJobs = j;
+	}
+}
+
+
+static void
+may_adjust_jobs_goal()
+{
+	char buffer[1024];
+	int filled = 0;
+	ssize_t n;
+	long value;
+	char *end;
+	int timeout;
+
+	/* basically, we can't escape with allocated == 0: first time through
+	 * we do it non-block. If we end up with empty allocated, we block !
+	 * (or rather: we set timeout to a high value so we can still 
+	 * see signals */
+	timeout = 100;
+	do {
+		/* since the messages are rather short, we assume a full 
+		 * message will always make it through.
+		 */
+		while ((n = poll(pollarray, 1, timeout)) == 1) {
+			if (pollarray[0].revents & POLLHUP) {
+				/* either the server went belly up or we didn't
+				 * identify properly */
+				goto error;
+			} else {
+				/* this is a bit tricky: we may have slept a 
+				 * long time, so there may be several events 
+				 * waiting for us -> keep polling/parsing job 
+				 * number lines. */
+				n = read(control_point, buffer+filled, 
+				    sizeof buffer-1-filled);
+				if (n == -1)
+					goto error;
+				filled += n;
+				buffer[filled] = 0;
+				while (filled > 0) {
+					value = strtol(buffer, &end, 10);
+					if (value < 0)
+						goto error;
+					if (*end == 0)
+						break; /* no full line */
+					/* skips thru terminators */
+					while (*end == '\r' || *end == '\n')
+						end++;
+					/* adjust goals temporarily */
+					jobs_goal = value;
+					/* and go for another round */
+					memmove(buffer, end, 
+					    sizeof buffer - (end-buffer));
+					filled -= (end-buffer);
+				}
+			}
+			timeout = 100;
+			continue;
+	error:
+			close(control_point);
+			control_point = -1;
+			break;
+		}
+		adjust_up();
+		timeout = INFTIM;
+	} while (allocated == 0);
+}
+
+/* various types of socket: if anything goes wrong, we just don't connect
+ * and we won't be able to adjust anything
+ */
+static int
+connect_to_local_addr(const char *name)
+{
+	struct sockaddr_un addr;
+	int s;
+
+	addr.sun_len = sizeof(addr);
+	addr.sun_family = AF_UNIX;
+	strlcpy(addr.sun_path, name, sizeof(addr.sun_path));
+
+	s = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (s == -1) {
+		return -1;
+	}
+	if (connect(s, (const struct sockaddr *)&addr, sizeof(addr)) == -1) {
+		close(s);
+		return -1;
+    	}
+	return s;
+}
+
+static int
+connect_to_inet(const char *server, const char *service)
+{
+	struct addrinfo hints, *res, *res0;
+	int error;
+	int s;
+
+	memset(&hints, 0, sizeof hints);
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+
+	error = getaddrinfo(server, service, &hints, &res0);
+	if (error)
+		return -1;
+
+	for (res = res0; res; res = res->ai_next) {
+		s = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+		if (s == -1)
+			continue;
+		if (connect(s, res->ai_addr, res->ai_addrlen) == -1) {
+			close(s);
+			s = -1;
+			continue;
+		}
+		break;
+	}
+	return s;
+}
+
+static int
+connect_to(char *addr)
+{
+	char *pos;
+
+	if (strchr(addr, '/')) {
+		return connect_to_local_addr(addr);
+	} else if ((pos = strchr(addr, ':')) != NULL) {
+		*pos = 0;
+		return connect_to_inet(addr, pos+1);
+	} else {
+		return connect_to_inet(addr, NULL);
+	}
+}
+
+void
+may_connect_to_controller(void)
+{
+	char *addr = getenv("BUILDCONTROLADDR");
+	char *token = getenv("BUILDTOKEN");
+	if (addr != NULL && strcmp(addr, "") != 0
+	    && token != NULL && strcmp(token, "") != 0) {
+		control_point = connect_to(addr);
+		if (control_point != -1) {
+			/* if we managed to connect, identify ourselves
+			 * to the builder.
+			 */
+			write(control_point, token, strlen(token));
+			/* Follow the full protocol, make it a full line */
+			write(control_point, "\r\n", 2);
+			pollarray[0].fd = control_point;
+			pollarray[0].events = POLLIN | POLLHUP;
+		}
+	}
 }
Index: job.h
===================================================================
RCS file: /build/data/openbsd/cvs/src/usr.bin/make/job.h,v
diff -u -p -r1.39 job.h
--- job.h	21 May 2024 05:00:48 -0000	1.39
+++ job.h	30 Jan 2025 14:34:38 -0000
@@ -96,7 +96,6 @@ extern void debug_job_printf(const char 
 extern int check_dying_signal(void);
 
 extern const char *basedirectory;
-
-extern bool	sequential;	/* True if we are running one single-job */
+extern void may_connect_to_controller(void);
 
 #endif /* _JOB_H_ */
Index: main.c
===================================================================
RCS file: /build/data/openbsd/cvs/src/usr.bin/make/main.c,v
diff -u -p -r1.133 main.c
--- main.c	18 Jun 2024 02:11:03 -0000	1.133
+++ main.c	30 Jan 2025 14:34:38 -0000
@@ -706,6 +706,7 @@ main(int argc, char **argv)
 	Var_Set("MACHINE", machine);
 	Var_Set("MACHINE_ARCH", machine_arch);
 	Var_Set("MACHINE_CPU", machine_cpu);
+	may_connect_to_controller();
 
 	/*
 	 * First snag any flags out of the MAKEFLAGS environment variable.
